src/background/responders.ts

```diff
@@ -130,5 +93,105 @@ export const responders: { [T in Action['type']]: Responder<T> } = {
       justTweeted: null,
       alert: 'POST TWEET FAILURE'
     }
+  },
+  'chrome.windows.getAll'(state, action) {
+    return {
+      focusedWindowId: action.payload.windows.find(win => win.focused)!.id
+    }
+  },
+  'chrome.tabs.query'(state, action) {
+    const tabs: AppState['tabs'] = new Map()
+    action.payload.tabs.forEach(tab =>
+      tabs.set(tab.id!, {
+        id: tab.id!,
+        windowId: tab.windowId!,
+        active: tab.active!,
+        url: tab.url!,
+        host: new URL(tab.url!).host,
+        feedbackState: {
+          screenshots: [],
+          editorState: EditorState.createEmpty()
+        }
+      })
+    )
+    return { tabs }
+  },
+  'chrome.tabs.onCreated'(state, action) {
+    const { tab } = action.payload
+    const tabs = new Map(state.tabs)
+    tabs.set(tab.id!, {
+      id: tab.id!,
+      windowId: tab.windowId!,
+      active: tab.active,
+      url: tab.url,
+      host: tab.url ? new URL(tab.url).host : undefined,
+      feedbackState: {
+        screenshots: [],
+        editorState: EditorState.createEmpty()
+      }
+    })
+    return { tabs }
+  },
+  'chrome.tabs.onRemoved'(state, action) {
+    const tabs = new Map(state.tabs)
+    tabs.delete(action.payload.tabId)
+    return { tabs }
+  },
+  'chrome.tabs.onUpdated'(state, action) {
+    const { tabId, changeInfo } = action.payload
+    if (!changeInfo.url) return {}
+    const tabs = new Map(state.tabs)
+    const tab = { ...tabs.get(tabId)! }
+    tab.url = changeInfo.url
+    tab.host = new URL(changeInfo.url).host
+    tabs.set(tabId, tab)
+    return { tabs }
+  },
+  'chrome.tabs.onAttached'(state, action) {
+    const tabs = new Map(state.tabs)
+    const {
+      tabId,
+      attachInfo: { newWindowId }
+    } = action.payload
+    const tab = { ...tabs.get(tabId)! }
+    tab.windowId = newWindowId
+    tabs.set(tabId, tab)
+    return { tabs }
+  },
+  'chrome.tabs.onActivated'(state, action) {
+    const tabs: AppState['tabs'] = new Map()
+    const { tabId, windowId } = action.payload.activeInfo
+
+    for (const [id, tab] of state.tabs.entries()) {
+      const nextTab = tab.windowId === windowId ? { ...tab, active: id === tabId } : tab
+      tabs.set(id, nextTab)
+    }
+
+    return { tabs }
+  },
+  'chrome.tabs.onReplaced'(state, action) {
+    console.log('chrome.tabs.onReplaced', state, action)
+    return {}
+  },
+  'chrome.windows.onCreated'(state, action) {
+    if (action.payload.win.focused) {
+      return { focusedWindowId: action.payload.win.id }
+    }
+    return {}
+  },
+  'chrome.windows.onRemoved'(state, action) {
+    const { windowId } = action.payload
+    const tabs = new Map(state.tabs)
+    for (const [id, tab] of state.tabs.entries()) {
+      if (tab.windowId === windowId) {
+        tabs.delete(id)
+      }
+    }
```

Comment:
It looks like `chrome.tabs.onRemoved` is being called before this event.

Reply:
re: `onReplaced` - how should we respond? The only data available is 
```
integer | addedTabId
integer | removedTabId
```

Reply:
As `chrome.tabs.onRemoved` being called before this event I don't mind covering our ass in case firefox/edge respond differently (I bet they don't, but this defense in depth can't hurt)

Reply:
create a copy of the removed `tab` object with the `addedTabId` as `id` and delete the tab `removedTabId`